### ISSUE TITLE
Ideas for PRs

### ISSUE THREAD

**Issue by willemneal** [AssemblyScript/meta#12]
- Add Template for the PRs
- All PRs start as drafts.
- If a user makes PR for their first time, a former contributor is picked at random (or from a team corresponding to that PR) and they will be the point person who can help approve the draft so that it becomes ready for others to see.  By building in mentor/mentee relationships, new users won't be as shy when asking questions.
- Currently the default writing permission of a PR's from branch is given to upstream contributors.  Thus when the draft is sent to the mentor to review, they can checkout the branch and push any corrections.  This can help new users because it will make it feel more like they were part of a team.
- When submitting a feature deemed ready for merging, the name of the feature can be added behind a flag in a nightly branch that users can test it on.
- Build reputation through PRs to join a team.

**Comment by MaxGraey** [comment]
- All PRs start as drafts.
Not sure about this. What about if PR is just minor fix or typo correction? Making it draft is unnecessary. Draft need in pretty rare case when you need some early feedback for huge & complex PR

But rest make sense

**Comment by willemneal** [comment]
It's related to your seniority.  If you are on a particular team you are authorized to make a PR however you like, though still with a simple template in mind, e.i. different templates for type of PR).

The nightly branch will start as draft pull request and others make PR's to it so that they will make up the changelog of the version.  However, minor fixes can go straight to master.

We could also set up a core team who are randomly selected from to review and approve changes to this branch.  Perhaps two or three are chosen and with enough approvals is merged automatically.  Then after milestones are met the nightly PR is approved and published automatically.

**Comment by MaxGraey** [comment]
How measure maturity? Count number merged (accepted) PRs? I guess it something overcomplicated. Not sure we need this

**Comment by dcodeIO** [comment]
> Add Template for the PRs

Sure, why not. But please let's not go overboard with it and consider it a guideline rather than a bureaucratic process.

> All PRs start as drafts.

Afaik this isn't supported out of the box by GitHub and a contributor would have to do this manually. We can add a sentence to the contributing guidelines that we'd prefer it this way ofc.

> If a user makes PR for their first time, a former contributor is picked at random (or from a team corresponding to that PR) and they will be the point person who can help approve the draft [...]

> Currently the default writing permission of a PR's from branch is given to upstream contributors. Thus when the draft is sent to the mentor to review, they can checkout the branch and push any corrections.

The idea is great, though I have my reservations that it'll work in the long run. We can't just assign someone at random since the person might not actually be interested or have the time (currently). Overall this seems hard to organize / sustain on our side. Maybe make that an external program driven by those interested in teaching?

> When submitting a feature deemed ready for merging, the name of the feature can be added behind a flag in a nightly branch that users can test it on.

It's rather rare that someone implements an entire feature that doesn't interfere with other parts of the code so we can easily put it behind a flag. Not sure if this is practical, if I understood it correctly. Once dist files aren't part of the repo anymore, an alternative way might be to simply npm install a specific GH branch in such cases.

> Build reputation through PRs to join a team.

My general feeling is that we cannot strictly measure this in numbers. Ideally, everyone who contributed in the past, no matter the extent, would be eligible to join a team upon request, since any disagreement on whether someone "deserves" it or not would be detrimental for all sides.

**Comment by torch2424** [comment]
I agree with the points made by @MaxGraey and @dcodeIO here. But, I definitely think it'd be helpful to have templates for PRs 😄 👍

**Comment by willemneal** [comment]
Having a precise measure is not the point, but rather some predetermined action such as making a PR to the project of the team once and then can apply to join the team in question.

I've been looking at github actions and it seems that we can use the action to authenticate the user and then have access to normal github APIs, which I'm looking at to see if we can change the state of the PR as part of the action triggered by the PR.

The idea behind assignment is that it would be from the team your on.  Those who want to help in the community by being mentors can join a teaching team.  And then have teams for various aspects, where assignments for review would also pull from when a PR is marked ready for review.

It doesn't have to be formal, just thinking of ways to review the code effectively as the community grows.
